File: commons/javafx/src/main/java/com/cedarsoft/commons/javafx/EnumListCell.java
package com.cedarsoft.commons.javafx;

import javax.annotation.Nonnull;

import javafx.scene.control.ComboBox;
import javafx.scene.control.ListCell;

/**
 * List cell that renders an enum
 */
public class EnumListCell<T extends Enum<T>> extends ListCell<T> {
  @Override
  protected void updateItem(T item, boolean empty) {
    super.updateItem(item, empty);
    if (item != null) {
      setText(EnumTranslatorUtil.getEnumTranslator().translate(item));
    }
  }

  public void apply(@Nonnull ComboBox<T> comboBox) {
    comboBox.setButtonCell(this);
    comboBox.setCellFactory(param -> new EnumListCell<>());
  }

  public static <T extends Enum<T>> void createFor(@Nonnull ComboBox<T> comboBox) {
    new EnumListCell<T>().apply(comboBox);
  }


}

File: commons/javafx/src/main/java/com/cedarsoft/commons/javafx/EnumListCell.kt
package com.cedarsoft.commons.javafx

import com.cedarsoft.commons.javafx.EnumTranslatorUtil.enumTranslator
import javafx.scene.control.ComboBox
import javafx.scene.control.ListCell
import javafx.scene.control.ListView
import javafx.util.Callback
import javax.annotation.Nonnull

/**
 * List cell that renders an enum
 */
class EnumListCell<T : Enum<T>> : ListCell<T>() {
  override fun updateItem(item: T?, empty: Boolean) {
    super.updateItem(item, empty)
    if (item != null) {
      text = enumTranslator.translate(item)
    }
  }

  fun apply(@Nonnull comboBox: ComboBox<T>) {
    comboBox.buttonCell = this
    comboBox.cellFactory = Callback { param: ListView<T>? -> EnumListCell() }
  }

  companion object {
    @JvmStatic
    fun <T : Enum<T>> createFor(@Nonnull comboBox: ComboBox<T>?) {
      EnumListCell<T>().apply(comboBox!!)
    }
  }
}
